File: cpp/neuralnet/coremlmodel.m
#import "coremlmodel.h"

@implementation KataGoModelInput

- (instancetype)initWithInput_spatial:(MLMultiArray *)input_spatial input_global:(MLMultiArray *)input_global {
  self = [super init];
  if (self) {
    _input_spatial = input_spatial;
    _input_global = input_global;
  }
  return self;
}

- (NSSet<NSString *> *)featureNames {
  return [NSSet setWithArray:@[@"input_spatial", @"input_global"]];
}

- (nullable MLFeatureValue *)featureValueForName:(NSString *)featureName {
  if ([featureName isEqualToString:@"input_spatial"]) {
    return [MLFeatureValue featureValueWithMultiArray:_input_spatial];
  }
  if ([featureName isEqualToString:@"input_global"]) {
    return [MLFeatureValue featureValueWithMultiArray:_input_global];
  }
  return nil;
}

@end

@implementation KataGoModelOutput

- (instancetype)initWithOutput_policy:(MLMultiArray *)output_policy out_value:(MLMultiArray *)out_value out_miscvalue:(MLMultiArray *)out_miscvalue out_moremiscvalue:(MLMultiArray *)out_moremiscvalue out_ownership:(MLMultiArray *)out_ownership {
  self = [super init];
  if (self) {
    _output_policy = output_policy;
    _out_value = out_value;
    _out_miscvalue = out_miscvalue;
    _out_moremiscvalue = out_moremiscvalue;
    _out_ownership = out_ownership;
  }
  return self;
}

- (NSSet<NSString *> *)featureNames {
  return [NSSet setWithArray:@[@"output_policy", @"out_value", @"out_miscvalue", @"out_moremiscvalue", @"out_ownership"]];
}

- (nullable MLFeatureValue *)featureValueForName:(NSString *)featureName {
  if ([featureName isEqualToString:@"output_policy"]) {
    return [MLFeatureValue featureValueWithMultiArray:_output_policy];
  }
  if ([featureName isEqualToString:@"out_value"]) {
    return [MLFeatureValue featureValueWithMultiArray:_out_value];
  }
  if ([featureName isEqualToString:@"out_miscvalue"]) {
    return [MLFeatureValue featureValueWithMultiArray:_out_miscvalue];
  }
  if ([featureName isEqualToString:@"out_moremiscvalue"]) {
    return [MLFeatureValue featureValueWithMultiArray:_out_moremiscvalue];
  }
  if ([featureName isEqualToString:@"out_ownership"]) {
    return [MLFeatureValue featureValueWithMultiArray:_out_ownership];
  }
  return nil;
}

@end

@implementation KataGoModel


/// Get URL of the MLModel at Application Support Directory.
/// - Parameters:
///   - modelName: The name of the MLModel.
+ (nullable NSURL *)getAppMLModelURL:(NSString * _Nonnull)modelName {
  // Get model package name
  NSString *mlpackageName = [NSString stringWithFormat:@"%@.mlpackage", modelName];

  // Set the directory for KataGo models
  NSString *directory = @"KataGoModels";

  // Get path component
  NSString *pathComponent = [NSString stringWithFormat:@"%@/%@", directory, mlpackageName];

  // Get default file manager
  NSFileManager *fileManager = [NSFileManager defaultManager];

  // Get application support directory
  // Create the directory if it does not already exist
  NSURL *appSupportURL = [fileManager URLForDirectory:NSApplicationSupportDirectory
                                             inDomain:NSUserDomainMask
                                    appropriateForURL:nil
                                               create:true
                                                error:nil];

  // Create the URL for the model package file
  NSURL *modelURL = [appSupportURL URLByAppendingPathComponent:pathComponent];

  return modelURL;
}


/// Compile the MLModel at Application Support Directory for KataGoModel and returns the compiled model.
/// - Parameters:
///   - modelName: The name of the MLModel.
+ (nullable MLModel *)compileAppMLModelWithModelName:(NSString * _Nonnull)modelName {

  // Get URL of the MLModel at Application Support Directory
  NSURL *modelURL = [KataGoModel getAppMLModelURL:modelName];

  // Check the MLModel is reachable
  BOOL isReachable = [modelURL checkResourceIsReachableAndReturnError:nil];

  MLModel *mlmodel = nil;

  if (isReachable) {
    // Compile MLModel if the MLModel is reachable
    mlmodel = [KataGoModel compileMLModelWithModelName:modelName
                                              modelURL:modelURL];
  }

  return mlmodel;
}


/// Compile the MLModel at bundle for KataGoModel and returns the compiled model.
/// - Parameters:
///   - modelName: The name of the MLModel.
+ (nullable MLModel *)compileBundleMLModelWithModelName:(NSString * _Nonnull)modelName {

  // Set model type name
  NSString *typeName = @"mlpackage";

  NSString *modelPath;

  // Get model path from bundle resource
  modelPath = [[NSBundle mainBundle] pathForResource:modelName
                                              ofType:typeName];

  if (modelPath == nil) {
    // Fallback to create a default model path
    modelPath = [NSString stringWithFormat:@"%@.%@", modelName, typeName];
  }

  // Get model URL at bundle
  NSURL *bundleModelURL = [NSURL fileURLWithPath:modelPath];

  // Compile MLModel
  MLModel *mlmodel = [KataGoModel compileMLModelWithModelName:modelName
                                                     modelURL:bundleModelURL];

  if (mlmodel != nil) {
    // Get model URL at App Support Directory
    NSURL *appModelURL = [KataGoModel getAppMLModelURL:modelName];

    // Get default file manager
    NSFileManager *fileManager = [NSFileManager defaultManager];

    NSLog(@"INFO: Removing old CoreML model in Application Support directory %@", appModelURL);

    // Remove the old model in Application Support directory
    [fileManager removeItemAtURL:appModelURL
                           error:nil];

    NSLog(@"INFO: Copying bundle CoreML model to Application Support directory %@", appModelURL);

    // Copy the mlpackage to App Support Directory
    BOOL success = [fileManager copyItemAtURL:bundleModelURL
                                        toURL:appModelURL
                                        error:nil];

    assert(success);
  }

  return mlmodel;
}

/// Compile the MLModel for KataGoModel and returns the compiled model.
/// - Parameters:
///   - modelName: The name of the MLModel.
///   - modelURL: The URL of the MLModel.
+ (nullable MLModel *)compileMLModelWithModelName:(NSString * _Nonnull)modelName
                                         modelURL:(NSURL * _Nonnull)modelURL {

  // Get compiled model name
  NSString *compiledModelName = [NSString stringWithFormat:@"%@.mlmodelc", modelName];

  // Set the directory for KataGo models
  NSString *directory = @"KataGoModels";

  // Get path component
  NSString *pathComponent = [NSString stringWithFormat:@"%@/%@", directory, compiledModelName];

  // Get default file manager
  NSFileManager *fileManager = [NSFileManager defaultManager];

  // Get application support directory
  // Create the directory if it does not already exist
  NSURL *appSupportURL = [fileManager URLForDirectory:NSApplicationSupportDirectory
                                             inDomain:NSUserDomainMask
                                    appropriateForURL:nil
                                               create:true
                                                error:nil];

  // Create the URL for the permanent compiled model file
  NSURL *permanentURL = [appSupportURL URLByAppendingPathComponent:pathComponent];

  // Initialize model
  MLModel *model = nil;

  // Create the URL for the model data file
  NSURL *dataURL = [modelURL URLByAppendingPathComponent:@"Data/com.apple.CoreML/model.mlmodel"];

  // Get model data
  NSData *modelData = [NSData dataWithContentsOfURL:dataURL];

  assert(modelData != nil);

  // Initialize hash data
  NSMutableData *hashData = [NSMutableData dataWithLength:CC_SHA256_DIGEST_LENGTH];

  // Get SHA256 data
  CC_SHA256(modelData.bytes, (CC_LONG)modelData.length, hashData.mutableBytes);

  // Get hash digest
  NSString *digest = [hashData base64EncodedStringWithOptions:0];

  // Set digest path
  NSString *savedDigestPath = [NSString stringWithFormat:@"%@/%@.digest", directory, modelName];

  // Get digest URL
  NSURL *savedDigestURL = [appSupportURL URLByAppendingPathComponent:savedDigestPath];

  // Get saved digest
  NSString *savedDigest = [NSString stringWithContentsOfURL:savedDigestURL encoding:NSUTF8StringEncoding error:nil];

  // Check permanent compiled model is reachable
  BOOL reachableModel = [permanentURL checkResourceIsReachableAndReturnError:nil];

  if (!reachableModel) {
    NSLog(@"INFO: Compiling CoreML model because it is not reachable");
  }

  // Check the saved digest is changed or not
  BOOL isChangedDigest = ![digest isEqualToString:savedDigest];

  if (isChangedDigest) {
    NSLog(@"INFO: Compiling CoreML model because the digest has changed");
  }

  // Model should be compiled if the compiled model is not reachable or the digest changes
  BOOL shouldCompile = !reachableModel || isChangedDigest;

  if (shouldCompile) {
    NSLog(@"INFO: Compiling CoreML model at %@", modelURL);

    // Compile the model
    NSURL *compiledURL = [MLModel compileModelAtURL:modelURL
                                              error:nil];

    NSLog(@"INFO: Copying the compiled CoreML model to the permanent location %@", permanentURL);

    // Create the directory for KataGo models
    BOOL success = [fileManager createDirectoryAtURL:[appSupportURL URLByAppendingPathComponent:directory]
                         withIntermediateDirectories:true
                                          attributes:nil
                                               error:nil];

    assert(success);

    // Copy the file to the to the permanent location, replacing it if necessary
    success = [fileManager replaceItemAtURL:permanentURL
                              withItemAtURL:compiledURL
                             backupItemName:nil
                                    options:NSFileManagerItemReplacementUsingNewMetadataOnly
                           resultingItemURL:nil
                                      error:nil];

    assert(success);

    // Update the digest
    success = [digest writeToURL:savedDigestURL
                      atomically:YES
                        encoding:NSUTF8StringEncoding
                           error:nil];

    assert(success);
  }

  // Initialize the model configuration
  MLModelConfiguration *configuration = [[MLModelConfiguration alloc] init];

  // Set the compute units to CPU and Neural Engine
  configuration.computeUnits = MLComputeUnitsCPUAndNeuralEngine;

  // Set the model display name
  configuration.modelDisplayName = modelName;

  NSLog(@"INFO: Creating CoreML model with contents %@", permanentURL);

  // Create the model
  model = [MLModel modelWithContentsOfURL:permanentURL
                            configuration:configuration
                                    error:nil];

  assert(model != nil);

  NSLog(@"INFO: Created CoreML model: %@", model.modelDescription.metadata[MLModelDescriptionKey]);

  // Return the model
  return model;
}


/**
 URL of the underlying .mlmodelc directory.
 */
+ (nullable NSURL *)URLOfModelInThisBundle {
  NSString *assetPath = [[NSBundle bundleForClass:[self class]] pathForResource:@"KataGoModel" ofType:@"mlmodelc"];
  if (nil == assetPath) { os_log_error(OS_LOG_DEFAULT, "Could not load KataGoModel.mlmodelc in the bundle resource"); return nil; }
  return [NSURL fileURLWithPath:assetPath];
}


/**
 Initialize KataGoModel instance from an existing MLModel object.

 Usually the application does not use this initializer unless it makes a subclass of KataGoModel.
 Such application may want to use `-[MLModel initWithContentsOfURL:configuration:error:]` and `+URLOfModelInThisBundle` to create a MLModel object to pass-in.
 */
- (instancetype)initWithMLModel:(MLModel *)model {
  self = [super init];
  if (!self) { return nil; }
  _model = model;
  if (_model == nil) { return nil; }
  return self;
}


/**
 Initialize KataGoModel instance with the model in this bundle.
 */
- (nullable instancetype)init {
  return [self initWithContentsOfURL:(NSURL * _Nonnull)self.class.URLOfModelInThisBundle error:nil];
}


/**
 Initialize KataGoModel instance from the model URL.

 @param modelURL URL to the .mlmodelc directory for KataGoModel.
 @param error If an error occurs, upon return contains an NSError object that describes the problem. If you are not interested in possible errors, pass in NULL.
 */
- (nullable instancetype)initWithContentsOfURL:(NSURL *)modelURL error:(NSError * _Nullable __autoreleasing * _Nullable)error {
  MLModel *model = [MLModel modelWithContentsOfURL:modelURL error:error];
  if (model == nil) { return nil; }
  return [self initWithMLModel:model];
}


/**
 Initialize KataGoModel instance from the model URL.

 @param modelURL URL to the .mlmodelc directory for KataGoModel.
 @param configuration The model configuration object
 @param error If an error occurs, upon return contains an NSError object that describes the problem. If you are not interested in possible errors, pass in NULL.
 */
- (nullable instancetype)initWithContentsOfURL:(NSURL *)modelURL configuration:(MLModelConfiguration *)configuration error:(NSError * _Nullable __autoreleasing * _Nullable)error {
  MLModel *model = [MLModel modelWithContentsOfURL:modelURL configuration:configuration error:error];
  if (model == nil) { return nil; }
  return [self initWithMLModel:model];
}

- (nullable KataGoModelOutput *)predictionFromFeatures:(KataGoModelInput *)input options:(MLPredictionOptions *)options error:(NSError * _Nullable __autoreleasing * _Nullable)error {
  id<MLFeatureProvider> outFeatures = [_model predictionFromFeatures:input options:options error:error];
  if (!outFeatures) { return nil; }
  return [[KataGoModelOutput alloc] initWithOutput_policy:(MLMultiArray *)[outFeatures featureValueForName:@"output_policy"].multiArrayValue out_value:(MLMultiArray *)[outFeatures featureValueForName:@"out_value"].multiArrayValue out_miscvalue:(MLMultiArray *)[outFeatures featureValueForName:@"out_miscvalue"].multiArrayValue out_moremiscvalue:(MLMultiArray *)[outFeatures featureValueForName:@"out_moremiscvalue"].multiArrayValue out_ownership:(MLMultiArray *)[outFeatures featureValueForName:@"out_ownership"].multiArrayValue];
}

@end

File: cpp/neuralnet/coremlmodel.swift
//
//  coremlmodel.swift
//  KataGo
//
//  Created by Chin-Chang Yang on 2023/11/7.
//

import CryptoKit
import Foundation
import CoreML

class KataGoModelInput: MLFeatureProvider {
    var input_spatial: MLMultiArray
    var input_global: MLMultiArray
    var input_meta: MLMultiArray?

    var featureNames: Set<String> {
        return Set(["input_spatial", "input_global", "input_meta"])
    }

    init(input_spatial: MLMultiArray, input_global: MLMultiArray) {
        self.input_spatial = input_spatial
        self.input_global = input_global
    }

    init(input_spatial: MLMultiArray, input_global: MLMultiArray, input_meta: MLMultiArray) {
        self.input_spatial = input_spatial
        self.input_global = input_global
        self.input_meta = input_meta
    }

    func featureValue(for featureName: String) -> MLFeatureValue? {
        if (featureName == "input_spatial") {
            return MLFeatureValue(multiArray: input_spatial)
        } else if (featureName == "input_global") {
            return MLFeatureValue(multiArray: input_global)
        } else if (featureName == "input_meta"), let input_meta {
            return MLFeatureValue(multiArray: input_meta)
        } else {
            return nil
        }
    }
}

class KataGoModelInputBatch: MLBatchProvider {
    var inputArray: [KataGoModelInput]

    var count: Int {
        inputArray.count
    }

    func features(at index: Int) -> MLFeatureProvider {
        return inputArray[index]
    }

    init(inputArray: [KataGoModelInput]) {
        self.inputArray = inputArray
    }
}

class KataGoModelOutput {
    var output_policy: MLMultiArray
    var out_value: MLMultiArray
    var out_miscvalue: MLMultiArray
    var out_moremiscvalue: MLMultiArray
    var out_ownership: MLMultiArray

    init(output_policy: MLMultiArray,
         out_value: MLMultiArray,
         out_miscvalue: MLMultiArray,
         out_moremiscvalue: MLMultiArray,
         out_ownership: MLMultiArray) {
        self.output_policy = output_policy
        self.out_value = out_value
        self.out_miscvalue = out_miscvalue
        self.out_moremiscvalue = out_moremiscvalue
        self.out_ownership = out_ownership
    }
}

class KataGoModelOutputBatch {
    var outputArray: [KataGoModelOutput]

    var count: Int {
        outputArray.count
    }

    init(outputArray: [KataGoModelOutput]) {
        self.outputArray = outputArray
    }
}

class KataGoModel {
    let model: MLModel

    class func getBundleModelURL(modelName: String) -> URL {
        // Set model type name
        let typeName = "mlpackage"
        // Get model path from bundle resource
        // Fallback to create a default model path
        let modelPath = Bundle.main.path(forResource: modelName, ofType: typeName) ?? "\(modelName).\(typeName)"
        let bundleModelURL = URL(filePath: modelPath)

        return bundleModelURL
    }

    class func compileBundleMLModel(modelName: String, computeUnits: MLComputeUnits, mustCompile: Bool = false) -> MLModel? {
        var mlmodel: MLModel?

        do {
            // Get model URL at bundle
            let bundleModelURL = getBundleModelURL(modelName: modelName)

            // Compile MLModel
            mlmodel = try compileMLModel(modelName: modelName,
                                         modelURL: bundleModelURL,
                                         computeUnits: computeUnits,
                                         mustCompile: mustCompile)
        } catch {
            printError("An error occurred: \(error)")
        }

        return mlmodel;
    }

    private class func getApplicationSupportURL() throws -> URL {
        // Get default file manager
        let fileManager = FileManager.default

        return try fileManager.url(for: .applicationSupportDirectory,
                                   in: .userDomainMask,
                                   appropriateFor: nil,
                                   create: true)
    }

    private class func getDigest(modelURL: URL) throws -> String {
        // Create the URL for the model data file
        let dataURL = modelURL.appending(component: "Data/com.apple.CoreML/model.mlmodel")

        // Get model data
        let modelData = try Data(contentsOf: dataURL)

        // Get SHA256 data
        let hashData = Data(SHA256.hash(data: modelData).makeIterator())

        // Get hash digest
        let digest = hashData.map { String(format: "%02x", $0) }.joined()

        return digest
    }

    private class func checkShouldCompileModel(permanentURL: URL,
                                               savedDigestURL: URL,
                                               digest: String) -> Bool {
        // Model should be compiled if the compiled model is not reachable or the digest changes
        var shouldCompile = true

        // Get saved digest
        do {
            if (try savedDigestURL.checkResourceIsReachable()) {
                let savedDigest = try String(contentsOf: savedDigestURL, encoding: .utf8)

                // Check the saved digest is changed or not
                shouldCompile = digest != savedDigest

                if (shouldCompile) {
                    printError("Saved digest: \(savedDigest)")
                    printError("New digest: \(digest)")
                    printError("Compiling CoreML model because the digest has changed");
                }
            } else {
                printError("Compiling CoreML model because the saved digest URL is not reachable: \(savedDigestURL)")
            }
        } catch {
            printError("Compiling CoreML model because it is unable to get the saved digest from: \(savedDigestURL)")
        }

        if !shouldCompile {
            // Check permanent compiled model is reachable
            do {
                // This method is currently applicable only to URLs for file system
                // resources. For other URL types, `false` is returned.
                shouldCompile = try (!permanentURL.checkResourceIsReachable())
                assert(!shouldCompile)
            } catch {
                shouldCompile = true

                printError("Compiling CoreML model because it is unable to check the resource at: \(permanentURL)")
            }
        }

        return shouldCompile
    }

    private class func compileAndSaveModel(permanentURL: URL,
                                           savedDigestURL: URL,
                                           modelURL: URL,
                                           digest: String) throws {
        // Get default file manager
        let fileManager = FileManager.default

        printError("Compiling CoreML model at \(modelURL)");

        // Compile the model
        let compiledURL = try MLModel.compileModel(at: modelURL)

        printError("Creating the directory for the permanent location: \(permanentURL)");

        // Create the directory for KataGo models
        try fileManager.createDirectory(at: permanentURL.deletingLastPathComponent(),
                                        withIntermediateDirectories: true)

        printError("Copying the compiled CoreML model to the permanent location \(permanentURL)");

        // Copy the file to the to the permanent location, replacing it if necessary
        try fileManager.replaceItem(at: permanentURL,
                                    withItemAt: compiledURL,
                                    backupItemName: nil,
                                    options: .usingNewMetadataOnly,
                                    resultingItemURL: nil)

        printError("Writing digest to: \(savedDigestURL)")
        printError("Digest: \(digest)")

        // Update the digest
        try digest.write(to: savedDigestURL, atomically: true, encoding: .utf8)
    }

    private class func loadModel(permanentURL: URL, modelName: String, computeUnits: MLComputeUnits) throws -> MLModel {
        let configuration = MLModelConfiguration()
        configuration.computeUnits = computeUnits
        configuration.modelDisplayName = modelName
        printError("Creating CoreML model with contents \(permanentURL)")
        return try MLModel(contentsOf: permanentURL, configuration: configuration)
    }

    class func getMLModelCPermanentURL(modelName: String) throws -> URL {
        let appSupportURL = try getApplicationSupportURL()
        let permanentURL = appSupportURL.appending(component: "KataGoModels/\(modelName).mlmodelc")

        return permanentURL
    }

    class func getSavedDigestURL(modelName: String) throws -> URL {
        let appSupportURL = try getApplicationSupportURL()
        let savedDigestURL = appSupportURL.appending(component: "KataGoModels/\(modelName).digest")

        return savedDigestURL
    }

    class func compileMLModel(modelName: String, modelURL: URL, computeUnits: MLComputeUnits, mustCompile: Bool) throws -> MLModel {
        let permanentURL = try getMLModelCPermanentURL(modelName: modelName)
        let savedDigestURL = try getSavedDigestURL(modelName: modelName)
        let digest = try getDigest(modelURL: modelURL)

        let shouldCompileModel = mustCompile || checkShouldCompileModel(permanentURL: permanentURL,
                                                                        savedDigestURL: savedDigestURL,
                                                                        digest: digest)

        if shouldCompileModel {
            try compileAndSaveModel(permanentURL: permanentURL,
                                    savedDigestURL: savedDigestURL,
                                    modelURL: modelURL,
                                    digest: digest)
        }

        return try loadModel(permanentURL: permanentURL,
                             modelName: modelName,
                             computeUnits: computeUnits);
    }

    init(model: MLModel) {
        self.model = model
    }

    private func createOutput(from outFeatures: MLFeatureProvider) -> KataGoModelOutput {

        let output_policy = (outFeatures.featureValue(for: "output_policy")?.multiArrayValue)!
        let out_value = (outFeatures.featureValue(for: "out_value")?.multiArrayValue)!
        let out_miscvalue = (outFeatures.featureValue(for: "out_miscvalue")?.multiArrayValue)!
        let out_moremiscvalue = (outFeatures.featureValue(for: "out_moremiscvalue")?.multiArrayValue)!
        let out_ownership = (outFeatures.featureValue(for: "out_ownership")?.multiArrayValue)!

        return KataGoModelOutput(output_policy: output_policy,
                                 out_value: out_value,
                                 out_miscvalue: out_miscvalue,
                                 out_moremiscvalue: out_moremiscvalue,
                                 out_ownership: out_ownership)
    }

    func prediction(from inputBatch: KataGoModelInputBatch,
                    options: MLPredictionOptions) throws -> KataGoModelOutputBatch {

        let outFeaturesBatch = try model.predictions(from: inputBatch, options: options)
        let outputArray = (0..<outFeaturesBatch.count).map { index -> KataGoModelOutput in
            let outFeatures = outFeaturesBatch.features(at: index)
            return createOutput(from: outFeatures)
        }

        return KataGoModelOutputBatch(outputArray: outputArray)
    }
}
